### ISSUE TITLE
Implement Gramian based aggregators as such

### ISSUE THREAD

**Issue by PierreQuinton** [TorchJD/torchjd#136]
This concerns:

- UPGRAD
- MGDA
- DualProj
- PCGrad
- IMTL-G
- CAGrad
- Nash-MTL
- Aligned-MTL

This is a strict improvement and is a first step towards Gramian based JD.

**Comment by ValerianRey** [comment]
Yes, for that we would need a third class hierarchy (the first is Aggregator, the second is _Weighting) that would be something like _GramianWeighting.

It would have the responsibility of giving weights (vector of shape [m]) from a Gramian (matrix of shape [m, m]).

I think I'd like to work on statefulness first, as it will also affect class hierarchies.

**Comment by ValerianRey** [comment]
I'm actually not sure about what to do for stateful aggregators, so we will probably do this first.
